File: node.js/lib/presenceChannel.js
var env = require('./env');
var Members = require('./members');
var Member = require('./member');

module.exports = PresenceChannel;

function PresenceChannel(name, appKey, cerrio,memberId,memberData,mainChannel, isPrivate){
    this.uri = 'PicTacToe/XStreamly/Presence';
    this.appKey = appKey;
    this.channelName = name;
    this.members = new Members();
    this.loaded = false;
    this.key = null;
    PresenceChannel.Instances= {};
    PresenceChannel.Instances[this.channelName] = this;
    var channelName = this.channelName;
    var timerId;

    //if no member id is defined then generate one that will be
    //constant across reconects
    if (!memberId) {
        if (!PresenceChannel.memberId) {
            PresenceChannel.memberId=Math.round(Math.random() * 1000000);
        }

        memberId = PresenceChannel.memberId.toString();
    }
    this.memberId = memberId;
    if(!memberData) {
      memberData = { memberId: memberId }
    }
    else if(!memberData.memberId){
    	memberData.memberId = memberId;
    }
    
    var me = this;

    this._memberInfo = memberData;

    this.memberInfo = function(name, value) {
      if(undefined === value && typeof name === 'string') {
        return me._memberInfo[name];
      } else {
        if(typeof name === 'string') {
          me._memberInfo[name] = value;
        } else if(typeof name === 'object') {
          $.each(name, function(key, value) {
            me._memberInfo[key] = value;
          });
        }
        cerrio.modify(uri, {
            Key: me.key,
            MemberInfo: JSON.stringify(me._memberInfo)
        });
      }
    }

    this.addAction =function(item){
      if(undefined === item.MemberId && item.MemberInfo) {
        try {
          item.MemberId = JSON.parse(item.MemberInfo).memberId;
        } catch(ex) { }
      }
    	var member = me.members.get(item.MemberId);
        if(undefined === member) {
            member = new Member(item.MemberId,item.MemberInfo);
            me.members.add(member);
        }
        member.addRecord(item);
        
        if(me.loaded){
            mainChannel.fireEvent("xstreamly:member_added",member);
        }
        
        
        if(item.Key===me.key){
            env.log('joined presence channel with:'+me.members.count);
            me.loaded=true;
            me.fireTimer();
            timerId =setInterval(me.fireTimer,10*1000);
            mainChannel.fireEvent("xstreamly:subscription_succeeded",me.members);
        }
    };
    
    this.deleteAction=function (item){
    	var member;
        me.members.each(function(m){
            if(m.containsKey(item.Key)){
                member=m;
            }
        });
        if(undefined!=member){
            member.removeRecord(item.Key);
            
            if(!member.alive()){
                me.members.remove(member.id);
                mainChannel.fireEvent("xstreamly:member_removed",member);
            }
        }
    };
    
    this.modifyAction=function(item){
    	var member;
        me.members.each(function(m){
            if(m.containsKey(item.Key)){
                member=m;
            }
        });
        if(undefined!=member){
          if(undefined!==item.MemberInfo) {
            try {
              member.memberInfo = JSON.parse(item.MemberInfo);
              mainChannel.fireEvent("xstreamly:member_modified",member);
            } catch(ex) { }
          }
        }
    };
       
    var stream = cerrio.subscribe({url:this.uri,
    	subscription:"@.AppKey='" + appKey + "' and @.Channel ='" + this.channelName + "' and @.Connected and @.Private ="+isPrivate,
    	addAction:this.addAction,
    	modifyAction:this.modifyAction,
    	deleteAction:this.deleteAction});

    this.fireTimer = function() {
	    cerrio.modify(uri, {
	        Key: me.key,
	        ChallengeTime: new Date(),
	        Connected: true
	    });
    };

    
    var fired = false;
    var uri = this.uri;
    var socket =this.socket;
    var action = function(){
        if(!fired){
        	me.key = appKey+'|'+channelName+'|'+cerrio.connection.socket.socket.sessionid+'|'+isPrivate;
            cerrio.add(uri,{
            	Key:me.key,
                Channel: channelName,
                AppKey:appKey,
                MemberId:memberId,
                MemberInfo: JSON.stringify(memberData),
                Verified: false,
                Connected: true,
                Challenge:0,
                ChallengeTime: (new Date()),
                Response:0,
                SocketId:cerrio.connection.socket.socket.sessionid,
                Private:isPrivate
            });
            //in cas the record already exists but is set to not connected
            //send in a mod            
           	cerrio.modify(uri, {
	        	Key: me.key,
	        	ChallengeTime: new Date(),
	        	Connected: true
	    	});
            fired=true;
        }
     };

     this.start = function() {
        env.log('starting prsence channel: '+this.channelName);
        cerrio.connection.onActive(action)
     };
     
     this.close = function(){
     	if(timerId){
     		clearInterval(timerId);
     		timerId = undefined;
     	}
     	
     	cerrio.sendDelete(uri,me.key);
     	
     	if(stream){
        	stream.close();
       	}
     }
}

File: java/src/ly/xstream/streaming/PresenceChannel.java
package ly.xstream.streaming;

import java.util.*;

import com.google.gson.reflect.TypeToken;

class PresenceChannel {
	private static String theMemberId;
	private static final String uri = "PicTacToe/XStreamly/Presence";
	
	private String appKey;
	private String channelName;
	private Boolean loaded = false;
	private String key;
	private Cerrio cerrio;
	private Boolean isPrivate;
	
	private Boolean startActionFired = false;
	
	public String memberId;
	public HashMap<String,String> memberInfo;
	
	private Stream stream;
	
	private Timer timer;
	
	private XstreamlyPresenceMessage message;
	
	Members members = new Members();
	
	private ILogger logger;
	
	private List<IXstreamlyChannelEventsHandler> channelEventsHandlers = new Vector<IXstreamlyChannelEventsHandler>();
	
	public PresenceChannel(final String name,final String appKey, final Cerrio cerrio, String memberId, HashMap<String,String> memberInfo, Boolean isPrivate,ILogger logger)
	{
		channelName = name;
		this.appKey = appKey;
		this.cerrio = cerrio;
		this.isPrivate=isPrivate;
		this.logger = logger;
		if(null==memberInfo){
			memberInfo = new HashMap<String, String>();
		}
		
		if(null==memberId){
			if(null==PresenceChannel.theMemberId){
				Random rand = new Random();
				PresenceChannel.theMemberId = Integer.toString(rand.nextInt(1000000));
			}
			this.memberId=PresenceChannel.theMemberId;
		} else {
			this.memberId = memberId;
		}
			
		memberInfo.put("memberId", this.memberId);
		
		this.memberInfo = memberInfo;
		
		SubscriptionOptions options = new SubscriptionOptions();
		options.subscription = "@.AppKey='" + appKey + "' and @.Channel ='" + this.channelName + "' and @.Connected and @.Private ="+isPrivate;
		options.uri=uri;
		options.addAction = new IDataUpdateHandler() {
			@Override
			public void handleUpdate(String data) {
				XstreamlyPresenceMessage message = StreamingClient.XStreaemlyGson.fromJson(data, XstreamlyPresenceMessage.class);
				addAction(message);

			}
		};
		options.modifyAction = new IDataUpdateHandler() {
			@Override
			public void handleUpdate(String data) {
				XstreamlyPresenceMessage message = StreamingClient.XStreaemlyGson.fromJson(data, XstreamlyPresenceMessage.class);
				modifyAction(message);
			}
		};
		options.deleteAction = new IDataUpdateHandler() {
			@Override
			public void handleUpdate(String data) {
				XstreamlyPresenceMessage message = StreamingClient.XStreaemlyGson.fromJson(data, XstreamlyPresenceMessage.class);
				deleteAction(message);
			}
		};
		
		timer = new Timer();
		timer.schedule(new TimerTask() {
			@Override
			public void run() {
				if(null!=message)
				{
					message.Connected = true;
					message.ChallengeTime = new Date();
					cerrio.sendModify(uri, message,"Connected","ChallengeTime","Key");
				}
				
			}
		},10*1000, 10*1000);
		
		stream = cerrio.subscribe(options);
	}
	
	public void start(){
		cerrio.connection.onActive(new IConnectionActiveCallback() {
			@Override
			public void onConnectionActive(Connection connection) {
				startAction();
			}
		}, false);
	}
	
	public void close()
	{
		if(null!=timer){
			timer.cancel();
			timer = null;
		}
		
		cerrio.sendDelete(uri, key);
		
		if(null!=stream){
			stream.close();
			stream = null;
		}
	}
	
	private void startAction(){
		if(startActionFired){
			return;
		}
		
		
		key = appKey+"|"+channelName+"|"+cerrio.connection.id+"|"+isPrivate;
		
		message = new XstreamlyPresenceMessage();
		
		message.memberId = this.memberId;
		message.memberInfo = this.memberInfo;
		
		message.Key=key;
		message.Channel=channelName;
		message.AppKey=appKey;
		message.MemberInfo=StreamingClient.XStreaemlyGson.toJson(memberInfo);
		message.Verified=false;
		message.Connected=true;
		message.Challenge=0;
		message.ChallengeTime=new Date();
		message.Response =0;
		message.SocketId = cerrio.connection.id;
		message.Private = false;
		
		cerrio.sendAdd(uri, message);
		cerrio.sendModify(uri, message);
		
		startActionFired = true;
	}
	
	public void bindToChannelEvents(IXstreamlyChannelEventsHandler handler){
			channelEventsHandlers.add(handler);
	}
	
	private void addAction(XstreamlyPresenceMessage message){
		if(null!= message.MemberInfo){
			message.memberInfo = StreamingClient.XStreaemlyGson.fromJson(message.MemberInfo,  new TypeToken<HashMap<String,String>>() {}.getType());
			message.memberId = message.memberInfo.get("memberId");
		}
		
		Member member = members.get(message.memberId);
		if(null==member){
			member = new Member(message.memberId,message.memberInfo);
			members.add(member);
		}
		
		member.addRecord(message);
		
		if(loaded){
			for(IXstreamlyChannelEventsHandler handler: channelEventsHandlers){
				handler.memberAdded(member);
			}
		}
		
		if(message.Key.equals(key)){
			logger.log("joined presence channel with: "+this.members.getCount());
			loaded = true;
			for(IXstreamlyChannelEventsHandler handler: channelEventsHandlers){
				handler.loaded(members);
			}
		}
	}
	
	private void modifyAction(XstreamlyPresenceMessage message){
		for(Member m: members.members.values()){
			if(m.containsKey(message.Key)){
				if(null!=message.MemberInfo){
					m.memberInfo = StreamingClient.XStreaemlyGson.fromJson(message.MemberInfo,  new TypeToken<HashMap<String,String>>() {}.getType());
					for(IXstreamlyChannelEventsHandler handler: channelEventsHandlers){
						handler.membmerModified(m);
					}
					return;
				}
			}
		}
	}

	private void deleteAction(XstreamlyPresenceMessage message){
		for(Member m: members.members.values()){
			if(m.containsKey(message.Key)){
				m.removeRecord(message.Key);
				if(!m.isAlive()){
					members.remove(m.id);
					for(IXstreamlyChannelEventsHandler handler: channelEventsHandlers){
						handler.memberRemoved(m);
					}
					return;
				}
			}
		}
	}
}
